qiskit/pulse/library/waveform.py

```diff
@@ -71,8 +72,11 @@ def _clip(self, samples: np.ndarray, epsilon: float = 1e-7) -> np.ndarray:
         Raises:
             PulseError: If there exists a pulse sample with a norm greater than 1+epsilon.
         """
+        # On some architectures `np.abs` results in floating-point errors
+        # of 1ULP, and the discrepancy needs to be accounted for.
+        ulp = float_info.epsilon
         samples_norm = np.abs(samples)
-        to_clip = (samples_norm > 1.0) & (samples_norm <= 1.0 + epsilon)
+        to_clip = (samples_norm > 1.0) & (samples_norm <= 1.0 + epsilon + ulp)
```

Comment:
I hope I've understood the question (and the originally intended underlying functionality) correctly.

My understanding is that `epsilon` is a user-facing parameter, enabling one to specify arbitrary precision tolerance.
The `ulp` has been added to account for the floating-point errors, which may result in going over the user-specified threshold by 1 ULP.

Somehow, it feels not right - although my intuition could be in an incorrect direction - to ask the user to account for the extra ULP by themselves, as part of the `epsilon` argument.

Reply:
It's fair to ask the user to also do it, but right now we're in the odd situation where we're passing things that look like they _should_ satisfy the constraints for `epsilon = 0` given how they're constructed, but are raising exceptions in very specialised cases if Numpy has access to certain SIMD instructions.  I think it's probably reasonable for us to increase the user-provided `epsilon` by one ULP to account for that.

Technically, the float_epsilon is only 1 ULP here if `0 <= epsilon < 1.0`, although really any epsilon outside that range is mostly nonsensical.  The most accurate change, since we have Numpy already imported, is to do the comparison to `samples_norm <= np.nextafter(1.0 + epsilon, np.inf)` - that will always raise the upper bound by 1 ULP, and the comment can just explain why that's necessary (if certain SIMD instructions are available, `np.abs` may disagree with `builtins.abs` by 1 ULP).

That just saves us from being in a case where the user says "look, `abs(x)` shows this number is `<= 1.0 + epsilon`, but Qiskit is still raising an exception!", since there's this weird minor discrepancy between the two FP libraries.